### PR TITLE
Task coordinator core business logic

### PR DESCRIPTION
**Pull request by AfanasievYuriy**

### CODE REVIEW

File: laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/service/impl/OrderServiceImpl.java [hunk 1]
```diff
@@ -98,6 +98,7 @@ public void completeNextOrderInQueue(int laundryId) {
                 TimeUnit.MILLISECONDS.sleep(orderEntity.getEstimatedTime());
                 break;
             } catch (InterruptedException ignore) {
+                //TODO shine2
```
Comment (andrsuh) [inline laundry-management-service/app/src/main/java/com/microservices/laundrymanagement/service/impl/OrderServiceImpl.java]: ?

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/controllers/CoordinatorController.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    private final LaundryStateService laundryStateService;
+
+    @Autowired
+    public CoordinatorController(OrderService orderService, LaundryStateService laundryStateService) {
+        this.orderService = orderService;
+        this.laundryStateService = laundryStateService;
+    }
+
+    @PostMapping()
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/controllers/CoordinatorController.java]: redundant ()

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/controllers/CoordinatorController.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    OrderSubmissionDTO coordinateOrder(@RequestBody OrderCoordinationDTO orderCoordinationDTO) {
+        return orderService.coordinateOrder(orderCoordinationDTO);
+    }
+
+    @PutMapping("/{id}/status/submitted")
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/controllers/CoordinatorController.java:29]: out docs that It is done for testing only

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/LaundryStateDTO.java [hunk 4]
```diff
@@ -1,17 +1,30 @@
...
-    int version;
+    private int laundryId;
+
+    private long queueWaitingTime;
+
+    private long reservedTime;
+
+    private int version;
+
+
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/LaundryStateDTO.java]: blank lines

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/OrderDetailDTO.java [hunk 5]
```diff
@@ -1,17 +1,31 @@
...
 
+import java.util.List;
+import java.util.stream.Collectors;
+
 @Getter
 @Setter
 @EqualsAndHashCode
 @AllArgsConstructor
+@NoArgsConstructor
 public class OrderDetailDTO {
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/OrderDetailDTO.java:19]: blank lines

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/inbound/OrderSubmittedDTO.java [hunk 6]
```diff
@@ -1,4 +1,16 @@
...
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+@AllArgsConstructor
+@EqualsAndHashCode
 public class OrderSubmittedDTO {
+
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/inbound/OrderSubmittedDTO.java:13]: blank

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/outbound/OrderSubmissionDTO.java [hunk 7]
```diff
@@ -1,17 +1,37 @@
...
+    public OrderSubmissionDTO(OrderEntity orderEntity, LaundryStateEntity laundryStateEntity) {
+        this.orderId = orderEntity.getId();
+        this.laundryId = orderEntity.getLaundryId();
+        this.bucket = orderEntity.getBucket();
+        this.orderDetails = orderEntity.getDetails().stream()
+                .map(OrderDetailDTO::new)
+                .collect(Collectors.toList());
+    }
+
+
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/outbound/OrderSubmissionDTO.java]: f* blanck lines

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/entity/OrderEntity.java [hunk 8]
```diff
@@ -37,14 +40,24 @@
 
     private long completionTime;
 
-    public OrderEntity(OrderDTO orderDTO, int selectedLaundryId, long estimatedTimeToComplete) {
-        this.id = orderDTO.getOrderId();
+    @OneToMany
+    @JoinColumn(name = "id", referencedColumnName = "id")
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/entity/OrderEntity.java]: the name attribute in enough.  The value should be orderId

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/repository/LaundryStateRepository.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+import com.microservices.taskcoordinator.entity.LaundryStateEntity;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.CrudRepository;
+
+public interface LaundryStateRepository extends CrudRepository<LaundryStateEntity, Integer> {
+
+    @Query(value = "select ls from LaundryStateEntity ls " +
+            "where (ls.queueWaitingTime + ls.reservedTime) = " +
+            "   (select max(ls2.queueWaitingTime + ls2.reservedTime) " +
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/repository/LaundryStateRepository.java]: max?

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/OrderService.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+import com.microservices.taskcoordinator.dto.OrderDTO;
+import com.microservices.taskcoordinator.dto.inbound.OrderCoordinationDTO;
+import com.microservices.taskcoordinator.dto.outbound.OrderSubmissionDTO;
+
+public interface OrderService {
+
+    OrderDTO updateOrder(OrderDTO order);
+
+    OrderDTO getOrderById(Integer id);
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/OrderService.java:11]: why Integer not int?

Comment (andrsuh) [reply taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/OrderService.java:11]: Ok, that might be

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/LaundryStateServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+        laundryStateEntity.setReservedTime(laundryStateEntity.getReservedTime() + orderDuration);
+
+        laundryStateRepository.save(laundryStateEntity);
+        return new LaundryStateDTO(laundryStateEntity);
+    }
+
+    @Override
+    @Transactional
+    public LaundryStateDTO updateLaundryStateWithOrderSubmitted(OrderSubmittedDTO orderSubmittedDTO) {
+        LaundryStateEntity laundryStateEntity = laundryStateRepository.findById(orderSubmittedDTO.getLaundryState().getLaundryId())
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/LaundryStateServiceImpl.java:47]: check null

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/LaundryStateServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+                .orElseThrow(() -> new IllegalArgumentException("LaundryState with id " + orderSubmittedDTO.getOrderId() + "doesn't exist"));
+        OrderDTO order = orderService.getOrderById(orderSubmittedDTO.getOrderId());
+
+        laundryStateEntity.setReservedTime(laundryStateEntity.getReservedTime() - order.getDuration());
+        laundryStateEntity.setQueueWaitingTime(orderSubmittedDTO.getLaundryState().getQueueWaitingTime());
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/LaundryStateServiceImpl.java]: this is updated even if received version is newer than existing

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/LaundryStateServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    @Override
+    @Transactional
+    public LaundryStateDTO updateLaundryStateWithOrderProcessed(OrderProcessedDTO orderProcessedDTO) {
+        LaundryStateEntity laundryStateEntity = laundryStateRepository.findById(orderProcessedDTO.getLaundryState().getLaundryId())
+                .orElseThrow(() -> new IllegalArgumentException("LaundryState with id " + orderProcessedDTO.getOrderId() + "doesn't exist"));
+        OrderDTO order = orderService.getOrderById(orderProcessedDTO.getOrderId());
+
+        laundryStateEntity.setQueueWaitingTime(orderProcessedDTO.getLaundryState().getQueueWaitingTime());
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/LaundryStateServiceImpl.java]: this is updated even if received version is newer than existing

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/OrderServiceImpl.java [hunk 14]
```diff
@@ -1,19 +1,72 @@
...
+@Service
 public class OrderServiceImpl implements OrderService {
 
-    @Autowired
-    private PredictionService predictionService;
+    private final OrderRepository orderRepository;
+
+    private final LaundryStateService laundryStateService;
+
+    private final PredictionService predictionService;
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/OrderServiceImpl.java]: final? it is inconsistent

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/OrderServiceImpl.java [hunk 15]
```diff
@@ -1,19 +1,72 @@
...
+    public OrderServiceImpl(OrderRepository orderRepository, LaundryStateService laundryStateService, PredictionService predictionService) {
+        this.orderRepository = orderRepository;
+        this.laundryStateService = laundryStateService;
+        this.predictionService = predictionService;
+    }
+
+    @Override
+    @Transactional
+    public OrderDTO updateOrder(OrderDTO orderDTO) {
+        OrderEntity existingOrder = orderRepository.findById(orderDTO.getId())
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/OrderServiceImpl.java:29]: check null

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/OrderServiceImpl.java [hunk 16]
```diff
@@ -1,19 +1,72 @@
...
+
+        return new OrderDTO(orderEntityFound);
+    }
 
     @Override
-    public OrderEntity submitOrder(OrderDTO orderDTO) {
-        return null;
+    @Transactional
+    public OrderSubmissionDTO coordinateOrder(OrderCoordinationDTO inboundOrder) {
+        if (orderRepository.existsById(inboundOrder.getOrderId()))
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/OrderServiceImpl.java]: {}

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/OrderServiceImpl.java [hunk 17]
```diff
@@ -1,19 +1,72 @@
...
+            throw new IllegalArgumentException("Order with id " + inboundOrder.getOrderId() + "already exists");
+
+        LaundryStateEntity leastLoadedLaundry = laundryStateService.getLeastLoadedLaundry();
+        long estimatedCompletionTime = predictionService.getOrderCompletionTimePrediction(leastLoadedLaundry);
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/OrderServiceImpl.java]: maybe that might be a single method?

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/PredictionService.java [hunk 18]
```diff
@@ -1,4 +1,12 @@
 package com.microservices.taskcoordinator.service.impl;
 
+import com.microservices.taskcoordinator.entity.LaundryStateEntity;
+import org.springframework.stereotype.Service;
+
+@Service
 public class PredictionService {
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/PredictionService.java]: no interface?

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/PredictionService.java [hunk 19]
```diff
@@ -1,4 +1,12 @@
...
+
+    public long getOrderCompletionTimePrediction(LaundryStateEntity laundryStateEntity) {
+        return laundryStateEntity.getQueueWaitingTime() + laundryStateEntity.getReservedTime();
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/PredictionService.java]: that might be simple the method of LaundryStateEntity

**Review by andrsuh — CHANGES_REQUESTED** [review]

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/controllers/CoordinatorController.java (hunk 3 above)
Comment (AfanasievYuriy) [reply taskcoordinator/src/main/java/com/microservices/taskcoordinator/controllers/CoordinatorController.java:29]: later
will create story for this

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/LaundryStateDTO.java (hunk 4 above)
Comment (AfanasievYuriy) [reply taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/LaundryStateDTO.java]: there is one line now

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/inbound/OrderSubmittedDTO.java (hunk 6 above)
Comment (AfanasievYuriy) [reply taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/inbound/OrderSubmittedDTO.java:13]: won't remove
by the way, Yulia P do the same thing in terms of code style

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/OrderService.java (hunk 10 above)
Comment (AfanasievYuriy) [reply taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/OrderService.java:11]: ? I guess that it has to be int not Integer. Why do u think Integer is ok?

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/OrderServiceImpl.java (hunk 17 above)
Comment (AfanasievYuriy) [reply taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/OrderServiceImpl.java]: might not

**Pull request closed without merging**